src/main/java/at/hannibal2/skyhanni/features/inventory/caketracker/CakeTracker.kt

```diff
@@ -0,0 +1,424 @@
+                ),
+            )
+        },
+    )
+
+    private fun drawDisplay(data: CakeData): List<Renderable> = buildList {
+        val dataHash = data.hashCode()
+        if (dataHash != lastKnownCakeDataHash) {
+            cakeRenderables = buildCakeRenderables(data)
+            lastKnownCakeDataHash = dataHash
+        }
+
+        addAll(cakeRenderables)
+    }
+
+    private fun buildCakeRenderables(data: CakeData) = buildList {
+        add(
+            Renderable.hoverTips(
+                "§c§lNew §f§lYear §c§lCake §f§lTracker",
+                tips = listOf("§aHave§7: §a${data.ownedCakes.count()}§7, §cMissing§7: §c${data.neededCakes.count()}"),
+            ),
+        )
+        add(buildDisplayTypeToggle())
+        add(buildOrderTypeToggle())
+
+        val cakeList = when (config.displayType) {
+            DisplayType.OWNED_CAKES -> data.ownedCakes
+            DisplayType.MISSING_CAKES -> data.neededCakes
+            null -> data.neededCakes
+        }
+
+        if (cakeList.isEmpty()) {
+            val colorCode = if (config.displayType == DisplayType.OWNED_CAKES) "§c" else "§a"
+            val verbiage = if (config.displayType == DisplayType.OWNED_CAKES) "missing" else "owned"
+            add(Renderable.string("$colorCode§lAll cakes $verbiage!"))
+        } else addCakeRanges(cakeList, config.displayOrderType, config.displayType)
+    }
+
+    private fun MutableList<Renderable>.addCakeRanges(
+        cakeList: List<Int>,
+        orderType: DisplayOrder,
+        displayType: DisplayType
+    ) {
+        val sortedCakes = when (orderType) {
+            DisplayOrder.OLDEST_FIRST -> cakeList.sorted()
+            DisplayOrder.NEWEST_FIRST -> cakeList.sortedDescending()
+        }.toMutableList()
+
+        // Combine consecutive years into ranges
+        val cakeRanges = mutableListOf<CakeRange>()
+        var start = sortedCakes.first()
+        var end = start
+
+        for (i in 1 until sortedCakes.size) {
+            if ((orderType == DisplayOrder.OLDEST_FIRST && sortedCakes[i] == end + 1) ||
+                (orderType == DisplayOrder.NEWEST_FIRST && sortedCakes[i] == end - 1)
+            ) {
+                end = sortedCakes[i]
+            } else {
+                if (start != end) cakeRanges.add(CakeRange(start, end))
+                else cakeRanges.add(CakeRange(start))
+                start = sortedCakes[i]
+                end = start
+            }
+        }
+
+        if (start != end) cakeRanges.add(CakeRange(start, end))
+        else cakeRanges.add(CakeRange(start))
+
+        // Store how many lines are 'hidden' as a result of the display limit
+        var hiddenRows = 0
+        cakeRanges.forEach {
+            // + 3 is to account for the header and selector boxes
+            if (this.size >= (config.maxDisplayRows.get() + 3)) hiddenRows++
+            else add(it.getRenderable(displayType))
+        }
```

Comment:
you dont need to iterate through it twice and generate renderable of lines that will be hidden later either way. if you want i can send a patch file in dms